lib/salesforce/database.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+require 'salesforce/donation_creator'
+require 'salesforce/supporter_creator'
+require 'salesforce/supporter_finder'
+
+module Salesforce
+  class Database
+    def self.add_donation(data)
+      new(data).add_donation
+    end
+
+    def initialize(data)
+      @data = data
+    end
+
+    def add_donation
+      return [:missing_email] unless email_present?
+      return supporter_result.errors unless supporter_result.okay?
+      donation_creation_errors
+    end
+
+    private
+
+    attr_reader :data
+
+    def email_present?
+      data.respond_to?(:email) && data.email
+    end
+
+    def supporter_result
+      @supporter_result ||= ensure_supporter_result
+    end
```

Comment:
because `@supporter_result ||= if problems_with_supporter_search? || supporter_exists?` is horribly unreadable and makes my eyes cry blood. Too much going on.